tracing-opentelemetry/src/layer.rs

```diff
@@ -0,0 +1,276 @@
+    let (trace_id, trace_flags) = builder
+        .parent_context
+        .as_ref()
+        .map(|parent_context| (parent_context.trace_id(), parent_context.trace_flags()))
+        .unwrap_or_else(|| {
+            (
+                builder.trace_id.expect("trace_id should exist"),
+                api::TRACE_FLAG_SAMPLED,
+            )
+        });
+
+    api::SpanContext::new(trace_id, span_id, trace_flags, false)
+}
+
+struct SpanEventVisitor<'a>(&'a mut api::Event);
+
+impl<'a> field::Visit for SpanEventVisitor<'a> {
+    /// Record events on the underlying OpenTelemetry `Span`.
+    fn record_debug(&mut self, field: &field::Field, value: &dyn fmt::Debug) {
+        if field.name() == "message" {
+            self.0.name = format!("{:?}", value);
+        } else {
+            self.0
+                .attributes
+                .push(api::Key::new(field.name()).string(format!("{:?}", value)));
+        }
+    }
+}
+
+struct SpanAttributeVisitor<'a>(&'a mut api::SpanBuilder);
+
+impl<'a> field::Visit for SpanAttributeVisitor<'a> {
+    /// Set attributes on the underlying OpenTelemetry `Span`.
+    fn record_debug(&mut self, field: &field::Field, value: &dyn fmt::Debug) {
+        let attribute = api::Key::new(field.name()).string(format!("{:?}", value));
+        if let Some(attributes) = &mut self.0.attributes {
+            attributes.push(attribute);
+        } else {
+            self.0.attributes = Some(vec![attribute]);
+        }
+    }
+}
+
+impl<S, T> OpenTelemetryLayer<S, T>
+where
+    S: Subscriber + for<'span> LookupSpan<'span>,
+    T: api::Tracer + 'static,
+{
+    /// Retrieve the parent OpenTelemetry [`SpanContext`] from the current
+    /// tracing [`span`] through the [`Registry`]. This [`SpanContext`]
+    /// links spans to their parent for proper hierarchical visualization.
+    fn parent_context(
+        &self,
+        attrs: &Attributes<'_>,
+        ctx: &Context<'_, S>,
+    ) -> Option<api::SpanContext> {
+        // If a span is specified, it _should_ exist in the underlying `Registry`.
+        if let Some(parent) = attrs.parent() {
+            let span = ctx.span(parent).expect("Span not found, this is a bug");
+            let mut extensions = span.extensions_mut();
+            extensions.get_mut::<api::SpanBuilder>().map(build_context)
+        // Else if the span is inferred from context, look up any available current span.
+        } else if attrs.is_contextual() {
+            ctx.current_span().id().and_then(|span_id| {
+                let span = ctx.span(span_id).expect("Span not found, this is a bug");
+                let mut extensions = span.extensions_mut();
+                extensions.get_mut::<api::SpanBuilder>().map(build_context)
+            })
+        // Explicit root spans should have no parent context.
+        } else {
+            None
+        }
+    }
+
+    /// Set the `OpenTelemetry` `Tracer` that this layer will use to produce
+    /// and track `Span`s.
```

Comment:
nit: would be nice if these links had refs eventually.

Reply:
yes!